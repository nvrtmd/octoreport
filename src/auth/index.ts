export * from './login';
export * from './userInfo';
export * from './token';
