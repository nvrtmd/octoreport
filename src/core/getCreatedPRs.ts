import { ActivityQueryOptions, PullRequest } from "../types";

export async function getCreatedPRs(
  options: ActivityQueryOptions
): Promise<PullRequest[]> {
  const prs: PullRequest[] = [];

  return prs;
}
