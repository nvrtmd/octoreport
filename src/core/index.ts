export * from './pr/combined';
export * from './pr/userCreated';
export * from './pr/userParticipated';
export * from './pr/filters';
export * from './utils';
