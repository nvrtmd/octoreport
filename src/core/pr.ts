import { fetchAllPRListInPeriod } from '../api';
import { PRQueryParams, PR } from '../types';

export function uniqueArray<T>(array: T[]): T[] {
  return Array.from(new Set(array));
}

export function filterPRListByTargetBranch(prList: PR[], targetBranch: string): PR[] {
  const parsedTargetBranch = targetBranch.includes(':') ? targetBranch.split(':')[1] : targetBranch;
  return prList.filter((pr) => pr.targetBranch === parsedTargetBranch);
}

async function getPRListByTargetBranch(options: PRQueryParams): Promise<PR[]> {
  const prList = await fetchAllPRListInPeriod(options);
  return filterPRListByTargetBranch(prList, options.targetBranch);
}

export async function getUserPRListByCreationAndParticipation(
  options: PRQueryParams,
): Promise<{ userCreatedPRList: PR[]; userParticipatedPRList: PR[] }> {
  const allPRList = await getPRListByTargetBranch(options);

  const createdPRList: PR[] = [];
  const participatedPRList: PR[] = [];

  allPRList.forEach((pr) => {
    if (pr.author === options.username) {
      createdPRList.push({
        ...pr,
        comments: uniqueArray(pr.comments ?? []),
        reviewers: uniqueArray(pr.reviewers),
      });
    } else if (pr.comments?.includes(options.username) || pr.reviewers.includes(options.username)) {
      participatedPRList.push({
        ...pr,
        comments: uniqueArray(pr.comments ?? []),
        reviewers: uniqueArray(pr.reviewers),
      });
    }
  });

  return { userCreatedPRList: createdPRList, userParticipatedPRList: participatedPRList };
}

export async function getUserCreatedPRListInPeriod(options: PRQueryParams): Promise<PR[]> {
  const allPRList = await getPRListByTargetBranch(options);

  return allPRList
    .filter((pr) => pr.author === options.username)
    .map((pr) => ({
      ...pr,
      comments: uniqueArray(pr.comments ?? []),
      reviewers: uniqueArray(pr.reviewers),
    }));
}

export async function getUserParticipatedPRListInPeriod(options: PRQueryParams): Promise<PR[]> {
  const allPRList = await getPRListByTargetBranch(options);

  const participatedPRList = allPRList
    .filter(
      (pr) => pr.comments?.includes(options.username) || pr.reviewers.includes(options.username),
    )
    .map((pr) => ({
      ...pr,
      comments: uniqueArray(pr.comments ?? []),
      reviewers: uniqueArray(pr.reviewers),
    }));

  return participatedPRList;
}

export async function getUserCreatedPRCountInPeriod(prList: PR[]): Promise<number> {
  return prList.length;
}

export async function getUserCreatedPRListInPeriodByLabel(
  prList: PR[],
  labelFilter: string[],
): Promise<PR[]> {
  return prList.filter((pr) =>
    labelFilter.some((filter) =>
      pr.labels.some((label) => label.toLowerCase().includes(filter.toLowerCase())),
    ),
  );
}

export async function getUserPRCountByLabelInPeriod(prList: PR[]): Promise<Record<string, number>> {
  const labelCountMap: Record<string, number> = {};
  prList.map((pr) => {
    pr.labels.forEach((label) => {
      labelCountMap[label] = (labelCountMap[label] || 0) + 1;
    });
  });
  return labelCountMap;
}
