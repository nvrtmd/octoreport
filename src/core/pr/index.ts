export * from './all';
export * from './common';
export * from './combined';
export * from './filters';
export * from './userCreated';
export * from './userParticipated';
