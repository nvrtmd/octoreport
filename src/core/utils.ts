import { DateTime } from 'luxon';

export function uniqueArray<T>(array: T[] | null): T[] | null {
  return array ? Array.from(new Set(array)) : null;
}

export function convertUTCISOToLocal(date: string): string {
  const timeZone = Intl.DateTimeFormat().resolvedOptions().timeZone;
  const local = DateTime.fromISO(date, { zone: 'utc' }).setZone(timeZone);
  return local.toFormat('yyyy-MM-dd HH:mm:ss');
}
