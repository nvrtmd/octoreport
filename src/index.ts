export * from './api';
export * from './core';
export * from './schemas';
export * from './types';
export * from './auth';
