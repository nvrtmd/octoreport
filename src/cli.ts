#!/usr/bin/env node
import inquirer from 'inquirer';

import { loginWithGitHubDeviceFlow } from './auth/login';
import { getToken, setToken } from './auth/token';
import { getUserPRListByCreationAndParticipation } from './core';

const GITHUB_CLIENT_ID = 'Ov23lia7pFpgs8ULT1DL';

if (process.argv[2] === 'login' || !getToken()) {
  const token = await loginWithGitHubDeviceFlow(GITHUB_CLIENT_ID);
  setToken(token);
  console.log('ðŸŽ‰ Authentication successful! You can now use octoreport.');
}

const answers = await inquirer.prompt([
  {
    type: 'input',
    name: 'username',
    message:
      'Optionally, enter the target GitHub username (e.g., octocat) (press Enter to skip and show PRs created by you):',
  },
  {
    type: 'input',
    name: 'repository',
    message: 'Please enter the repository in the format "owner/repo" (e.g., facebook/react):',
  },
  {
    type: 'input',
    name: 'startDate',
    message: 'Please enter the start date for the search period (format: YYYY-MM-DD):',
  },
  {
    type: 'input',
    name: 'endDate',
    message: 'Please enter the end date for the search period (format: YYYY-MM-DD):',
  },
  {
    type: 'input',
    name: 'targetBranch',
    message:
      'Optionally, enter the target branch to filter pull requests (press Enter to skip and show PRs targeting all branches):',
  },
]);

const { userCreatedPRList, userParticipatedPRList } = await getUserPRListByCreationAndParticipation(
  {
    githubToken: getToken() ?? '',
    username: answers.username,
    repository: answers.repository,
    period: {
      startDate: answers.startDate,
      endDate: answers.endDate,
    },
    targetBranch: answers.targetBranch,
  },
);

console.log('User Created PRs: ', userCreatedPRList);
console.log('User Participated PRs: ', userParticipatedPRList);
