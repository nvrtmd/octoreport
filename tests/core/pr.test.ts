import { describe, expect, it } from "vitest";
import { getUserCreatedPRListInPeriod } from "../../src/core";
import { config } from "dotenv";
config();

describe("getCreatedPRs", () => {
  it("returns an empty array when the user has no pull requests in the given period", async () => {
    const result = await getUserCreatedPRListInPeriod({
      githubToken: process.env.GITHUB_TOKEN ?? "",
      username: "gaearon",
      repository: "bluesky-social/social-app",
      period: { startDate: "2020-11-01", endDate: "2020-11-30" },
      targetBranch: "main",
    });

    expect(result).toEqual([]);
  });

  it("returns pull requests created by the user within the specified date range", async () => {
    const result = await getUserCreatedPRListInPeriod({
      githubToken: process.env.GITHUB_TOKEN ?? "",
      username: "gaearon",
      repository: "bluesky-social/social-app",
      period: { startDate: "2024-11-20", endDate: "2024-11-30" },
      targetBranch: "main",
    });

    expect(result).toMatchSnapshot();
  });

  it("filters pull requests that target a specific branch", async () => {
    const result = await getUserCreatedPRListInPeriod({
      githubToken: process.env.GITHUB_TOKEN ?? "",
      username: "zpao",
      repository: "facebook/react",
      period: { startDate: "2021-04-10", endDate: "2021-04-20" },
      targetBranch: "facebook:master",
    });

    expect(result).toMatchSnapshot();
  });
});
