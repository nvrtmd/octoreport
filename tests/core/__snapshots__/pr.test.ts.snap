// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getUserCreatedPRListInPeriod > filters pull requests that target a specific branch 1`] = `
[
  {
    "author": "oliviertassinari",
    "createdAt": "2024-09-11T12:44:25Z",
    "labels": [
      "scope: docs-infra",
      "security",
    ],
    "mergedAt": "2024-09-30T12:36:38Z",
    "number": 43711,
    "reviewers": [
      "oliviertassinari",
      "oliviertassinari",
    ],
    "title": "[docs-infra] Strengthen CSP",
    "url": "https://github.com/mui/material-ui/pull/43711",
  },
  {
    "author": "oliviertassinari",
    "createdAt": "2024-09-11T13:30:02Z",
    "labels": [
      "core",
    ],
    "mergedAt": "2024-09-11T20:39:34Z",
    "number": 43712,
    "reviewers": [],
    "title": "[core] Remove /.yarn",
    "url": "https://github.com/mui/material-ui/pull/43712",
  },
  {
    "author": "oliviertassinari",
    "createdAt": "2024-09-13T17:58:16Z",
    "labels": [
      "component: slider",
      "docs",
    ],
    "mergedAt": "2024-09-15T15:05:26Z",
    "number": 43748,
    "reviewers": [
      "aarongarciah",
    ],
    "title": "[slider][docs] Polish Music player demo",
    "url": "https://github.com/mui/material-ui/pull/43748",
  },
  {
    "author": "oliviertassinari",
    "createdAt": "2024-09-14T00:04:37Z",
    "labels": [
      "docs",
      "regression 🐛",
      "scope: docs-infra",
    ],
    "mergedAt": "2024-09-16T17:19:10Z",
    "number": 43751,
    "reviewers": [
      "mapache-salvaje",
    ],
    "title": "[docs-infra] Fix Vale config for TypeScript references",
    "url": "https://github.com/mui/material-ui/pull/43751",
  },
  {
    "author": "oliviertassinari",
    "createdAt": "2024-09-14T13:38:28Z",
    "labels": [
      "docs",
      "regression 🐛",
    ],
    "mergedAt": "2024-09-16T06:41:54Z",
    "number": 43752,
    "reviewers": [
      "siriwatknp",
    ],
    "title": "[docs] Fix MUI Treasury Layout broken links",
    "url": "https://github.com/mui/material-ui/pull/43752",
  },
  {
    "author": "oliviertassinari",
    "createdAt": "2024-09-14T13:43:03Z",
    "labels": [
      "docs",
      "regression 🐛",
      "v5.x",
    ],
    "mergedAt": "2024-09-16T11:17:55Z",
    "number": 43753,
    "reviewers": [
      "siriwatknp",
    ],
    "title": "[docs] Fix MUI Treasury Layout broken links",
    "url": "https://github.com/mui/material-ui/pull/43753",
  },
  {
    "author": "oliviertassinari",
    "createdAt": "2024-09-14T14:15:20Z",
    "labels": [
      "bug 🐛",
      "examples",
      "v5.x",
    ],
    "mergedAt": "2024-10-03T11:40:08Z",
    "number": 43755,
    "reviewers": [
      "mnajdova",
    ],
    "title": "[examples] Fix v5 clone example instructions",
    "url": "https://github.com/mui/material-ui/pull/43755",
  },
  {
    "author": "oliviertassinari",
    "createdAt": "2024-09-15T00:29:26Z",
    "labels": [
      "bug 🐛",
      "core",
    ],
    "mergedAt": "2024-09-17T13:00:19Z",
    "number": 43778,
    "reviewers": [
      "DiegoAndai",
    ],
    "title": "[core] useAutocomplete should have 'use client'",
    "url": "https://github.com/mui/material-ui/pull/43778",
  },
  {
    "author": "oliviertassinari",
    "createdAt": "2024-09-18T12:15:03Z",
    "labels": [
      "docs",
    ],
    "mergedAt": "2024-09-19T12:10:40Z",
    "number": 43809,
    "reviewers": [
      "michaldudak",
      "michaldudak",
      "oliviertassinari",
    ],
    "title": "[docs] Fix 301 to chromium",
    "url": "https://github.com/mui/material-ui/pull/43809",
  },
  {
    "author": "oliviertassinari",
    "createdAt": "2024-09-18T16:26:18Z",
    "labels": [
      "docs",
    ],
    "mergedAt": "2024-09-30T18:14:43Z",
    "number": 43812,
    "reviewers": [
      "aarongarciah",
    ],
    "title": "[docs] Format Pigment CSS docs",
    "url": "https://github.com/mui/material-ui/pull/43812",
  },
  {
    "author": "oliviertassinari",
    "createdAt": "2024-09-19T19:34:40Z",
    "labels": [
      "core",
      "test",
    ],
    "mergedAt": "2024-09-20T11:23:07Z",
    "number": 43820,
    "reviewers": [
      "aarongarciah",
      "mnajdova",
      "sai6855",
    ],
    "title": "[test] Sync React.version parse logic with codebase",
    "url": "https://github.com/mui/material-ui/pull/43820",
  },
  {
    "author": "oliviertassinari",
    "createdAt": "2024-09-19T22:49:09Z",
    "labels": [
      "website",
    ],
    "mergedAt": "2024-09-28T10:16:12Z",
    "number": 43822,
    "reviewers": [
      "aarongarciah",
    ],
    "title": "[website] Improve Next roles section",
    "url": "https://github.com/mui/material-ui/pull/43822",
  },
  {
    "author": "oliviertassinari",
    "createdAt": "2024-09-20T20:37:20Z",
    "labels": [
      "website",
    ],
    "mergedAt": "2024-09-21T00:19:36Z",
    "number": 43831,
    "reviewers": [
      "MBilalShafi",
    ],
    "title": "[website] Add 'Row spanning'",
    "url": "https://github.com/mui/material-ui/pull/43831",
  },
  {
    "author": "oliviertassinari",
    "createdAt": "2024-09-20T23:47:01Z",
    "labels": [
      "regression 🐛",
      "scope: docs-infra",
    ],
    "mergedAt": "2024-10-02T21:51:40Z",
    "number": 43834,
    "reviewers": [],
    "title": "[docs-infra] Fix link in header regression",
    "url": "https://github.com/mui/material-ui/pull/43834",
  },
]
`;

exports[`getUserCreatedPRListInPeriod > returns pull requests created by the user within the specified date range 1`] = `
[
  {
    "author": "oliviertassinari",
    "createdAt": "2025-05-01T00:27:00Z",
    "labels": [
      "MUI X",
      "core",
      "regression 🐛",
    ],
    "mergedAt": "2025-05-05T23:41:19Z",
    "number": 46051,
    "reviewers": [
      "Janpot",
      "Janpot",
      "jedwards1211",
      "jedwards1211",
      "oliviertassinari",
      "oliviertassinari",
      "oliviertassinari",
    ],
    "title": "[core] Fix internal version duplication",
    "url": "https://github.com/mui/material-ui/pull/46051",
  },
  {
    "author": "oliviertassinari",
    "createdAt": "2025-05-02T13:29:40Z",
    "labels": [
      "regression 🐛",
      "website",
    ],
    "mergedAt": "2025-05-06T15:03:19Z",
    "number": 46061,
    "reviewers": [
      "nadjakovacev",
    ],
    "title": "[website] Update values",
    "url": "https://github.com/mui/material-ui/pull/46061",
  },
  {
    "author": "oliviertassinari",
    "createdAt": "2025-05-03T14:05:12Z",
    "labels": [
      "enhancement",
      "scope: docs-infra",
    ],
    "mergedAt": "2025-05-05T17:27:16Z",
    "number": 46064,
    "reviewers": [
      "alexfauquette",
      "oliviertassinari",
      "oliviertassinari",
      "oliviertassinari",
    ],
    "title": "[docs-infra] Update feedback Node.js to v22",
    "url": "https://github.com/mui/material-ui/pull/46064",
  },
  {
    "author": "oliviertassinari",
    "createdAt": "2025-05-03T15:34:05Z",
    "labels": [
      "regression 🐛",
      "website",
    ],
    "mergedAt": "2025-05-07T17:54:29Z",
    "number": 46065,
    "reviewers": [
      "mapache-salvaje",
    ],
    "title": "[website] Fix CSS alignment",
    "url": "https://github.com/mui/material-ui/pull/46065",
  },
  {
    "author": "oliviertassinari",
    "createdAt": "2025-05-05T22:16:02Z",
    "labels": [
      "docs",
      "enhancement",
      "scope: docs-infra",
    ],
    "mergedAt": "2025-05-07T20:47:34Z",
    "number": 46090,
    "reviewers": [
      "mapache-salvaje",
    ],
    "title": "[docs] Remove oudated scaffoldhub ad",
    "url": "https://github.com/mui/material-ui/pull/46090",
  },
  {
    "author": "oliviertassinari",
    "createdAt": "2025-05-12T18:54:56Z",
    "labels": [
      "docs",
    ],
    "mergedAt": "2025-05-15T07:27:18Z",
    "number": 46135,
    "reviewers": [
      "0210shivam",
      "0210shivam",
      "ZeeshanTamboli",
      "ZeeshanTamboli",
    ],
    "title": "[docs] Fix GridLegacy docs order",
    "url": "https://github.com/mui/material-ui/pull/46135",
  },
]
`;

exports[`getUserCreatedPRListInPeriodByLabel > filters pull requests that exactly match provided label names (case-sensitive) 1`] = `
[
  {
    "author": "oliviertassinari",
    "createdAt": "2025-05-03T14:05:12Z",
    "labels": [
      "enhancement",
      "scope: docs-infra",
    ],
    "mergedAt": "2025-05-05T17:27:16Z",
    "number": 46064,
    "reviewers": [
      "alexfauquette",
      "oliviertassinari",
      "oliviertassinari",
      "oliviertassinari",
    ],
    "title": "[docs-infra] Update feedback Node.js to v22",
    "url": "https://github.com/mui/material-ui/pull/46064",
  },
  {
    "author": "oliviertassinari",
    "createdAt": "2025-05-05T22:16:02Z",
    "labels": [
      "docs",
      "enhancement",
      "scope: docs-infra",
    ],
    "mergedAt": "2025-05-07T20:47:34Z",
    "number": 46090,
    "reviewers": [
      "mapache-salvaje",
    ],
    "title": "[docs] Remove oudated scaffoldhub ad",
    "url": "https://github.com/mui/material-ui/pull/46090",
  },
  {
    "author": "oliviertassinari",
    "createdAt": "2025-05-12T18:54:56Z",
    "labels": [
      "docs",
    ],
    "mergedAt": "2025-05-15T07:27:18Z",
    "number": 46135,
    "reviewers": [
      "0210shivam",
      "0210shivam",
      "ZeeshanTamboli",
      "ZeeshanTamboli",
    ],
    "title": "[docs] Fix GridLegacy docs order",
    "url": "https://github.com/mui/material-ui/pull/46135",
  },
]
`;

exports[`getUserCreatedPRListInPeriodByLabel > filters pull requests that include provided label keywords (case-insensitive, partial match) 1`] = `
[
  {
    "author": "oliviertassinari",
    "createdAt": "2025-05-01T00:27:00Z",
    "labels": [
      "MUI X",
      "core",
      "regression 🐛",
    ],
    "mergedAt": "2025-05-05T23:41:19Z",
    "number": 46051,
    "reviewers": [
      "Janpot",
      "Janpot",
      "jedwards1211",
      "jedwards1211",
      "oliviertassinari",
      "oliviertassinari",
      "oliviertassinari",
    ],
    "title": "[core] Fix internal version duplication",
    "url": "https://github.com/mui/material-ui/pull/46051",
  },
  {
    "author": "oliviertassinari",
    "createdAt": "2025-05-02T13:29:40Z",
    "labels": [
      "regression 🐛",
      "website",
    ],
    "mergedAt": "2025-05-06T15:03:19Z",
    "number": 46061,
    "reviewers": [
      "nadjakovacev",
    ],
    "title": "[website] Update values",
    "url": "https://github.com/mui/material-ui/pull/46061",
  },
  {
    "author": "oliviertassinari",
    "createdAt": "2025-05-03T15:34:05Z",
    "labels": [
      "regression 🐛",
      "website",
    ],
    "mergedAt": "2025-05-07T17:54:29Z",
    "number": 46065,
    "reviewers": [
      "mapache-salvaje",
    ],
    "title": "[website] Fix CSS alignment",
    "url": "https://github.com/mui/material-ui/pull/46065",
  },
]
`;
